### ISSUE TITLE
FTP.ftp_STOR() should handle IsADirectoryError error

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6284]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)| @adiroiban reported|
|-|-|
|Trac ID|trac#6284|
|Type|defect|
|Created|2013-01-29 11:22:42Z|
|Branch|https://github.com/twisted/twisted/tree/ftp_STOR_isadirectory-6284|

This is the second part of [#5441](https://github.com/twisted/twisted/issues/5441)

Right now, the FileHandler raised a IsADirectory error when trying to write over an existing folder.

The FTP.ftp_STOR() handle IsNotADirectoryError so the IsADirectory error remains unhandled.

I will attach the diff

Attachments:

* [6248-1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1e4/1e4f45aec983b6e26f4eda228e05d4e16ce1e225/078d870fc40d86d9928facf391108b4d4d5fc30e.diff) (4110 bytes) - added by adiroiban on 2013-01-29 11:38:55Z - 
* [6284-2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1e4/1e4f45aec983b6e26f4eda228e05d4e16ce1e225/e8e76f1acd1cca4fe9b5d13b152a07fff1a5d0bc.diff) (1739 bytes) - added by adiroiban on 2013-03-26 22:35:07Z - applied on top of 7fd78ad tomprince 2013-03-18 Apply 6248-ftp-stor-folder.diff from adiroiban.

<details><summary>Searchable metadata</summary>

```
trac-id__6284 6284
type__defect defect
reporter__adiroiban adiroiban
priority__normal normal
milestone__ 
branch__branches_ftp_STOR_isadirectory_6284 branches/ftp_STOR_isadirectory-6284
branch_author__tomprince tomprince
status__closed closed
resolution__fixed fixed
component__ftp ftp
keywords__None None
time__1359458562000000 1359458562000000
changetime__1474198049187875 1474198049187875
version__None None
owner__tom_prince tom.prince
cc__itamarst
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_2"></a>|@adiroiban removed owner|
|-|-|

Hi,

I have attached the patch.

The error is due to a copy/paste from ftp_RMD .

I have also added docstring to callbed and tried to re-order them to improve reading the code.

There is also a live diff here:

https://github.com/chevah/twisted/compare/chevah:master...chevah:6248-ftp-stor-folder

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_3"></a>|@tomprince set owner to @adiroiban|
|-|-|

Thanks for your work on this.
  * `ebOpened` wants to log errors that aren't `FTPCmdError`, at least. It probably does want to keep the white-listing behaviour, since there are a bunch of `FTPCmdError`s that shouldn't be happening at that point (and thus are unexpected).
  * `ebSent` looking at the code that could lead to `ebSent`, it doesn't look like there is any expected error, so I think any error there should be logged.
  * Since you are changing how errors are handled in all he callbacks, you should add tests for all the cases. (coverage reports at least one with the current code)
Please resubmit after addressing the above points.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban set owner to @tomprince|
|-|-|

Hi and thanks for the review,

Replying to [tom.prince](#note_3):
> Thanks for your work on this.
>   * `ebOpened` wants to log errors that aren't `FTPCmdError`, at least. It probably does want to keep the white-listing behaviour, since there are a bunch of `FTPCmdError`s that shouldn't be happening at that point (and thus are unexpected).

I have added a log for non-FTPCmdError exceptions.

>   * `ebSent`: looking at the code that could lead to `ebSent`, it doesn't look like there is any expected error, so I think any error there should be logged.

The current DTP protocol implementation is very simple and it does not signal errors to the FTP protocol.

Custom implemtations of DTP protocol could raise a FTPCmdError instead of the current:
`raise Exception("Crap damn crap damn crap damn")`

https://github.com/chevah/twisted/blob/6248-ftp-stor-folder/twisted/protocols/ftp.py#L470

This is why I chose to not log FTPCmdError and they should not be considered "unexpected" errors.

The code from master for ebSent is just returning FTPCmdError, but I don't see where that returned error is handled. This is why I handle FTPCmdError in ebSent.

>   * Since you are changing how errors are handled in all he callbacks, you should add tests for all the cases. (coverage reports at least one with the current code)
> Please resubmit after addressing the above points.

I added coverage tests for error handling.

Nice diff is here: https://github.com/chevah/twisted/compare/chevah:master...chevah:6248-ftp-stor-folder

Can always be downloaded by appending .diff to the url

https://github.com/chevah/twisted/compare/chevah:master...chevah:6248-ftp-stor-folder.diff


Let me know if you want to to upload a new diff file.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_5"></a>|@tomprince set owner to @adiroiban|
|-|-|

>> * ebSent: looking at the code that could lead to ebSent, it doesn't look like there is any expected error, so I think any error there should be logged.
> The current DTP protocol implementation is very simple and it does not signal errors to the FTP protocol.
>
> Custom implemtations of DTP protocol could raise a FTPCmdError instead of the current: raise Exception("Crap damn crap damn crap damn") 

There is no way to use a custom dtp implementation. It is an implementation detail, so we don't need to worry about handling that case. Also, the errback is only going to get called if the transfer aborted, so that is the error we should return.

Otherwise, it looks good.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_6"></a>|@tomprince commented|
|-|-|

(In [37562]) Branching to ftp_STOR_isadirectory-6284.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban set owner to @tomprince|
|-|-|

t.w.p.FTP.dtpFactory is documented as a public method and it can be replaced with custom implementations.

Is there anyone using the default DTP implementation in production?!?!

With code like 'raise Exception("Crap damn crap damn crap damn")' this looks to me only like an example toy implementation.

Cheers,
Adi

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_8"></a>|@tomprince set owner to @adiroiban|
|-|-|

If you look at
https://github.com/tomprince/twisted/blob/trunk/twisted/protocols/ftp.py#L894
and
https://github.com/tomprince/twisted/blob/trunk/twisted/protocols/ftp.py#L894
which are  the two place where `dtpFactory` is assigned, doesn't give a user any place to override the factory before creating it. In any case, it isn't an interesting protocol, to implement a different way.

Regarding the exception, that code has been there from 2005. And, that error is not going to be propagated beyond `ebSent`.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban set owner to @tomprince|
|-|-|

Hi,

Thanks for the review and sorry for the trouble with my previous comment, in which I was wrong.

I have reverted ebSent implementation to the current code and adapted the new tests.

Diff is here:  https://github.com/chevah/twisted/compare/chevah:master...chevah:6248-ftp-stor-folder

Looks like GitHub got confused by this diff and it does not look to great. A side by side diff would help here.

Can always be downloaded by appending .diff to the url

https://github.com/chevah/twisted/compare/chevah:master...chevah:6248-ftp-stor-folder.diff

Let me know if you want it attached as a file.

Cheers,
Adi

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_10"></a>|@tomprince commented|
|-|-|

The linked patch doesn't apply against the branch.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_11"></a>|@adiroiban commented|
|-|-|

Hi. I have attached a new diff on top of the branch dedicated to this ticket.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_12"></a>|@tomprince commented|
|-|-|

(In [37942]) Apply 6284-2.diff from adiroiban.

Refs: [#6284](https://github.com/twisted/twisted/issues/6284)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_13"></a>|@tomprince set status to `closed`|
|-|-|

(In [37947]) Merge ftp_STOR_isadirectory-6284: Properly handle IsADirectoryError in ftp_STOR.

Author: adiroiban
Reviewers: tom.prince
Fixes: [#6284](https://github.com/twisted/twisted/issues/6284)

For some reason, it was handling IsNotADirectoryError, instead.
Now any FTPCmdError is passed through, rather than being logged.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_14"></a>|@hawkowl commented|
|-|-|

[mass edit] Removing review from closed tickets.
